### PR TITLE
More `rustc_infer` cleanups

### PR DESCRIPTION
**Pull request by nnethercote**
A sequel to #131226.

r? @lcnr

### CODE REVIEW

File: compiler/rustc_infer/src/infer/mod.rs [hunk 1]
```diff
@@ -68,6 +68,12 @@ pub mod resolve;
 pub(crate) mod snapshot;
 mod type_variable;
 
+// `InferOk<'tcx, ()>` is used a lot. It may seem like a useless wrapper around
+// `Vec<PredicateObligation<'tcx>>`, but it has one important property: because
+// `InferOk` is marked with `#[must_use]`, if you have a method `InferCtxt::f`
+// that returns `InferResult<'tcx, ()>` and you call it with `infcx.f()?;`
+// you'll get a warning about the obligations being discarded without use,
+// which is probably unintentional and has been a source of bugs in the past.
```
Comment (lcnr) [inline compiler/rustc_infer/src/infer/mod.rs:77]: doc comment?

File: compiler/rustc_infer/src/infer/region_constraints/leak_check.rs [hunk 2]
```diff
@@ -156,34 +157,24 @@ impl<'a, 'b, 'tcx> LeakCheck<'a, 'b, 'tcx> {
             // Detect those SCCs that directly contain a placeholder
             if let ty::RePlaceholder(placeholder) = **region {
                 if self.outer_universe.cannot_name(placeholder.universe) {
-                    self.assign_scc_value(scc, placeholder)?;
+                    // Update `scc_placeholders` to account for the fact that `P: S` must hold.
+                    // This may create an error.
```
Comment (lcnr) [inline compiler/rustc_infer/src/infer/region_constraints/leak_check.rs:161]:
```suggestion
```
that's obvious now, given that the `return Err` is inlined :thinking:

File: compiler/rustc_infer/src/infer/relate/combine.rs [hunk 3]
```diff
@@ -177,6 +179,8 @@ impl<'tcx> InferCtxt<'tcx> {
     where
         R: PredicateEmittingRelation<InferCtxt<'tcx>>,
     {
+        use InferConst::*;
```
Comment (lcnr) [inline compiler/rustc_infer/src/infer/relate/combine.rs:182]: personally not a big fan of this commit, but :shrug: I don't care much either way

**Comment by lcnr** [comment]
r=me after nits

File: compiler/rustc_infer/src/infer/mod.rs [hunk 4]
```diff
@@ -1037,18 +1044,17 @@ impl<'tcx> InferCtxt<'tcx> {
...
     pub fn get_region_var_origins(&self) -> VarInfos {
-        let mut inner = self.inner.borrow_mut();
-        let (var_infos, data) = inner
+        let inner = self.inner.borrow();
+        assert!(!UndoLogs::<UndoLog<'_>>::in_snapshot(&inner.undo_log));
+        let storage = inner
             .region_constraint_storage
             // We clone instead of taking because borrowck still wants to use
             // the inference context after calling this for diagnostics
             // and the new trait solver.
```
Comment (tjallingt) [inline compiler/rustc_infer/src/infer/mod.rs:1053]: I think this comment should be moved down to the `storage.var_infos.clone()`

File: compiler/rustc_infer/src/infer/relate/combine.rs (hunk 3 above)
Comment (compiler-errors) [reply compiler/rustc_infer/src/infer/relate/combine.rs:182]: I'd rather not use glob imports like this, so I'd prefer if we remove it.

**Comment by nnethercote** [comment]
I addressed the comments.

@bors r=lcnr

**Pull request merged**